README.md
# Luzifer / restis

`restis`, composed from `REST` and `Redis`, is a simple HTTP API to get / set / delete Redis keys.

## Q&A

- **Why?**  
  Previously I used a kyototycoon to store simple key-value-data through a REST interface but needed something not using a proprietary database format below it to have it running statelessly in a cluster. Searching for some alternative I gave up and invested 20 minutes in putting together this.
- **Is there any security built in?**  
  No. And it never will be. Don't expose it without securing it for example using [nginx-sso](https://github.com/Luzifer/nginx-sso) or Gatekeeper or whatever. This is just a simple API, you're responsible for the rest.

## Usage

```console
# restis --help
Usage of ./restis:
      --disable-cors               Disable setting CORS headers for all requests
      --listen string              Port/IP to listen on (default ":3000")
      --log-level string           Log level (debug, info, warn, error, fatal) (default "info")
      --redis-conn-string string   Connection string for redis (default "redis://localhost:6379/0")
      --redis-key-prefix string    Prefix to prepend to keys (will be prepended without delimiter!)
      --version                    Prints current version and exits
```

```console
# curl -fX GET localhost:3000/mykey
curl: (22) The requested URL returned error: 404

# echo "mycontent" | curl -fX PUT --data-binary @- localhost:3000/mykey

# curl -fX GET localhost:3000/mykey
mycontent

# curl -fX DELETE localhost:3000/mykey

# curl -fX GET localhost:3000/mykey
curl: (22) The requested URL returned error: 404

# echo "mycontent" | curl -fX PUT --data-binary @- localhost:3000/mykey\?expire=30s

# curl -fX GET localhost:3000/mykey
mycontent

# sleep 30; curl -fX GET localhost:3000/mykey
curl: (22) The requested URL returned error: 404
```


History.md
# 0.1.0 / 2022-12-28

  * Initial version


cors.go
package main

import (
	"net/http"
	"strings"
)

func corsMiddleware(h http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// Allow the client to send us credentials
		w.Header().Set("Access-Control-Allow-Credentials", "true")

		// We don't care about headers at all, allow sending all
		w.Header().Set("Access-Control-Allow-Headers", "*")

		// List all accepted methods no matter whether they are accepted by the specified endpoint
		w.Header().Set("Access-Control-Allow-Methods", strings.Join([]string{
			http.MethodDelete,
			http.MethodGet,
			http.MethodPut,
		}, ", "))

		// Public API: Let everyone in
		if origin := r.Header.Get("Origin"); origin != "" {
			w.Header().Set("Access-Control-Allow-Origin", origin)
		}

		h.ServeHTTP(w, r)
	})
}


rest.go
package main

import (
	"bytes"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/go-redis/redis/v8"
	"github.com/pkg/errors"
)

func keyFromRequest(r *http.Request) string {
	key := strings.TrimLeft(r.URL.Path, "/")
	if cfg.RedisKeyPrefix != "" {
		key = cfg.RedisKeyPrefix + key
	}
	return key
}

func handlerDelete(client *redis.Client) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if err := client.Del(r.Context(), keyFromRequest(r)).Err(); err != nil {
			http.Error(w, errors.Wrap(err, "deleting key").Error(), http.StatusInternalServerError)
			return
		}

		w.WriteHeader(http.StatusNoContent)
	}
}

func handlerGet(client *redis.Client) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		content, err := client.Get(r.Context(), keyFromRequest(r)).Bytes()
		switch {
		case err == nil:
			w.WriteHeader(http.StatusOK)
			w.Write(content)

		case errors.Is(err, redis.Nil):
			w.WriteHeader(http.StatusNotFound)

		default:
			http.Error(w, errors.Wrap(err, "getting key").Error(), http.StatusInternalServerError)
		}
	}
}

func handlerPut(client *redis.Client) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var (
			data   = new(bytes.Buffer)
			err    error
			expire time.Duration
		)

		if rawEx := r.URL.Query().Get("expire"); rawEx != "" {
			if expire, err = time.ParseDuration(rawEx); err != nil {
				http.Error(w, errors.Wrap(err, "parsing expiry").Error(), http.StatusBadRequest)
				return
			}
		}

		if _, err = io.Copy(data, r.Body); err != nil {
			http.Error(w, errors.Wrap(err, "reading payload").Error(), http.StatusBadRequest)
			return
		}

		if err = client.Set(r.Context(), keyFromRequest(r), data.Bytes(), expire).Err(); err != nil {
			http.Error(w, errors.Wrap(err, "setting key").Error(), http.StatusInternalServerError)
			return
		}

		w.WriteHeader(http.StatusCreated)
	}
}


main.go
package main

import (
	"fmt"
	"net/http"
	"os"
	"time"

	"github.com/go-redis/redis/v8"
	"github.com/gorilla/mux"
	"github.com/pkg/errors"
	"github.com/sirupsen/logrus"

	"github.com/Luzifer/rconfig/v2"
)

var (
	cfg = struct {
		DisableCORS     bool   `flag:"disable-cors" default:"false" description:"Disable setting CORS headers for all requests"`
		Listen          string `flag:"listen" default:":3000" description:"Port/IP to listen on"`
		LogLevel        string `flag:"log-level" default:"info" description:"Log level (debug, info, warn, error, fatal)"`
		RedisConnString string `flag:"redis-conn-string" default:"redis://localhost:6379/0" description:"Connection string for redis"`
		RedisKeyPrefix  string `flag:"redis-key-prefix" default:"" description:"Prefix to prepend to keys (will be prepended without delimiter!)"`
		VersionAndExit  bool   `flag:"version" default:"false" description:"Prints current version and exits"`
	}{}

	version = "dev"
)

func initApp() error {
	rconfig.AutoEnv(true)
	if err := rconfig.ParseAndValidate(&cfg); err != nil {
		return errors.Wrap(err, "parsing cli options")
	}

	if cfg.VersionAndExit {
		fmt.Printf("git-changerelease %s\n", version)
		os.Exit(0)
	}

	l, err := logrus.ParseLevel(cfg.LogLevel)
	if err != nil {
		return errors.Wrap(err, "parsing log-level")
	}
	logrus.SetLevel(l)

	return nil
}

func main() {
	var err error
	if err = initApp(); err != nil {
		logrus.WithError(err).Fatal("initializing app")
	}

	redisOpts, err := redis.ParseURL(cfg.RedisConnString)
	if err != nil {
		logrus.WithError(err).Fatal("parsing redis connection string")
	}

	var (
		client = redis.NewClient(redisOpts)
		router = mux.NewRouter()
	)

	if !cfg.DisableCORS {
		router.Use(corsMiddleware)
	}

	router.MethodNotAllowedHandler = corsMiddleware(http.HandlerFunc(func(res http.ResponseWriter, r *http.Request) {
		if r.Method == http.MethodOptions {
			// Most likely JS client asking for CORS headers
			res.WriteHeader(http.StatusNoContent)
			return
		}

		res.WriteHeader(http.StatusMethodNotAllowed)
	}))

	router.Methods(http.MethodDelete).HandlerFunc(handlerDelete(client))
	router.Methods(http.MethodGet).HandlerFunc(handlerGet(client))
	router.Methods(http.MethodPut).HandlerFunc(handlerPut(client))

	server := &http.Server{
		Addr:              cfg.Listen,
		Handler:           router,
		ReadHeaderTimeout: time.Second,
	}

	logrus.WithFields(logrus.Fields{
		"addr":    cfg.Listen,
		"version": version,
	}).Info("starting HTTP server")

	if err := server.ListenAndServe(); err != nil {
		logrus.WithError(err).Fatal("HTTP server quit unexpectedly")
	}
}
